python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1381 @@
+
+    def inspect(self):
+        cdef CResult[shared_ptr[CSchema]] result
+        with nogil:
+            result = self.discovery.Inspect()
+        return pyarrow_wrap_schema(GetResultValue(result))
+
+    def finish(self):
+        cdef CResult[shared_ptr[CDataSource]] result
+        with nogil:
+            result = self.discovery.Finish()
+        return DataSource.wrap(GetResultValue(result))
+
+
+cdef class FileSystemDataSourceDiscovery(DataSourceDiscovery):
+
+    cdef:
+        CFileSystemDataSourceDiscovery* filesystem_discovery
+
+    def __init__(self, FileSystem filesystem not None,
+                 Selector selector not None, FileFormat format not None,
+                 FileSystemDiscoveryOptions options=None):
+        cdef CResult[shared_ptr[CDataSourceDiscovery]] result
+        options = options or FileSystemDiscoveryOptions()
+        result = CFileSystemDataSourceDiscovery.Make(
+            filesystem.unwrap(),
+            selector.unwrap(),
+            format.unwrap(),
+            options.unwrap()
+        )
+        self.init(GetResultValue(result))
+
+    cdef init(self, shared_ptr[CDataSourceDiscovery]& sp):
+        DataSourceDiscovery.init(self, sp)
+        self.filesystem_discovery = <CFileSystemDataSourceDiscovery*> sp.get()
+
+
+cdef class FileSource:
+
+    cdef:
+        shared_ptr[CFileSource] wrapped
+        CFileSource* source
+
+    cdef readonly:
+        FileSystem fs
+
+    def __init__(self, path, FileSystem filesystem not None, compression=None):
+        cdef shared_ptr[CFileSource] source
+
+        # need to hold a reference for the filesystem
+        self.fs = filesystem
+
+        source.reset(new CFileSource(
+            tobytes(path),
+            self.fs.unwrap().get(),
+            _get_compression_type(compression)
+        ))
+
+        self.init(source)
+
+    cdef init(self, shared_ptr[CFileSource]& sp):
+        self.wrapped = sp
+        self.source = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CFileSource]& sp):
+        cdef FileSource self = FileSource.__new__(FileSource)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CFileSource] unwrap(self):
+        return self.wrapped
+
+    def equals(self, FileSource other):
+        return deref(self.source) == deref(other.unwrap())
+
+    def __eq__(self, other):
+        try:
+            return self.equals(other)
+        except TypeError:
+            return NotImplemented
+
+    @property
+    def compression(self):
+        return self.wrapped.get().compression()
+
+    @property
+    def path(self):
+        return frombytes(self.wrapped.get().path())
+
+    # def open(self):
+    #     CStatus Open(shared_ptr[CRandomAccessFile]* out)
+
+
+cdef class DataFragment:
+    """A granular piece of a Dataset such as an individual file.
+
+    It can be read/scanned separately from other fragments.
+    A DataFragment yields a collection of RecordBatch, encapsulated in one or
+    more ScanTasks.
```

Comment:
Maybe mention this is not directly created but is only created from `DataSource.fragments()`

Reply:
I've removed the fragments method.